src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs

```diff
@@ -22,18 +22,57 @@ public InstalledAndTransitivePackageFeed(IEnumerable<PackageCollectionItem> inst
         }
 
+        internal static PackageCollectionItem SelectTransitiveLatestPackage(IGrouping<string, PackageCollectionItem> groupedPackagesById, IEnumerable<PackageCollectionItem> installedPkgs)
+        {
+            if (groupedPackagesById == null)
+            {
+                throw new ArgumentNullException(nameof(groupedPackagesById));
+            }
+
+            if (installedPkgs == null)
+            {
+                throw new ArgumentNullException(nameof(installedPkgs));
+            }
+
+            IEnumerable<PackageCollectionItem> matchedOrigins = groupedPackagesById
+                .Where(x => x.PackageReferences
+                    .Cast<ITransitivePackageReferenceContextInfo>()
+                    .Any(trPr => trPr.TransitiveOrigins.Any(trOrigin => installedPkgs.Contains(trOrigin.Identity))));
+
+            if (matchedOrigins.Any())
+            {
+                // Return the latest transitive package with installed versions
+                return matchedOrigins.OrderByDescending(x => x.Version).First();
+            }
+
+            // Otherwise, fallback on what's availble
+            return groupedPackagesById.OrderByDescending(x => x.Version).First();
+        }
+
         /// <inheritdoc cref="IPackageFeed.ContinueSearchAsync(ContinuationToken, CancellationToken)" />
         public override async Task<SearchResult<IPackageSearchMetadata>> ContinueSearchAsync(ContinuationToken continuationToken, CancellationToken cancellationToken)
         {
             var searchToken = continuationToken as FeedSearchContinuationToken ?? throw new InvalidOperationException(Strings.Exception_InvalidContinuationToken);
             cancellationToken.ThrowIfCancellationRequested();
 
+            PackageCollectionItem[] installedFeedItems = _installedPackages.GetLatest();
+            PackageCollectionItem[] matchesInstalled = PerformLookup(installedFeedItems, searchToken);
+
             // Remove transitive packages from project references
             IEnumerable<PackageCollectionItem> pkgsWithOrigins = _transitivePackages
                 .Where(t => t.PackageReferences.Any(x => x is ITransitivePackageReferenceContextInfo y && y.TransitiveOrigins.Any()));
 
-            IPackageSearchMetadata[] installedItems = await GetMetadataForPackagesAndSortAsync(PerformLookup(_installedPackages, searchToken), searchToken.SearchFilter.IncludePrerelease, cancellationToken);
-            IPackageSearchMetadata[] transitiveItems = await GetMetadataForPackagesAndSortAsync(PerformLookup(pkgsWithOrigins, searchToken), searchToken.SearchFilter.IncludePrerelease, cancellationToken);
+            // Search all transitive packages, then group by package ID later
+            PackageCollectionItem[] matchesTransitive = PerformLookup(pkgsWithOrigins, searchToken);
+
+            // Select latest version with matching transitive origin with installed packages
+            // or fallback to latest version available
```

Comment:
I don't understand what this comment is telling us this code is supposed to be doing. Please see if we can explain this in different terms, making sure to be clear about the relationship between the version, the package, the transitive origin, and installed packages.

Reply:
I should have put this comment in line 72. I will rephrase.

Reply:
What do you think of:

```suggestion
matchesTransitive contains (packageId, version) PackageIdentity-like objects that represent matched transitive packages from previous search query.

In multi-targeting scenarion, it's common that multiple transitive packages can have the same ID, and we want to show only one package per package ID. Then, we need to select which transitive package version is best for showing in PM UI.

For each matched transitive package, we will either show:

- Latest transitive package version whose transitive origin matches one of latest installed package version, or
- show latest transitive package if none matches the above
```

Reply:
Rephrased comment.